diff --git a/server/process.c b/server/process.c
index a0d5ea64d97f1dcf80896081b750837c426a8e2f..ad1259fc10994a5db724f7f9b7ab46c4d5832eaf 100644
--- a/server/process.c
+++ b/server/process.c
@@ -861,7 +861,7 @@ static void process_poll_event( struct fd *fd, int event )
     struct process *process = get_fd_user( fd );
     assert( process->obj.ops == &process_ops );
 
-    if (event & (POLLERR | POLLHUP)) kill_process( process, 0 );
+    if (event & (POLLERR | POLLHUP)) kill_process( process, !process->is_terminating );
     else if (event & POLLIN) receive_fd( process );
 }
 
